File: spring-web-java/src/main/java/spring/web/java/application/event/NotificationEventListener.java
package spring.web.java.application.event;

import org.springframework.context.event.EventListener;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import spring.web.java.domain.model.entity.Member;
import spring.web.java.domain.model.entity.Notification;
import spring.web.java.domain.repository.MemberRepository;
import spring.web.java.domain.repository.NotificationRepository;

@Slf4j
@Component
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class NotificationEventListener {

	private final MemberRepository memberRepository;
	private final NotificationRepository notificationRepository;

	@Async
	@EventListener
	@Transactional
	public void handleNotificationEvent(NotificationEvent notificationEvent) {
		try {
			Member member = memberRepository.findById(notificationEvent.memberId()).orElseThrow(RuntimeException::new);

			notificationRepository.save(
				Notification.of(
					member,
					notificationEvent.title(),
					notificationEvent.message(),
					notificationEvent.url()
				)
			);
		} catch (Exception e) {
			log.error("알림 이벤트 처리 중 예외 발생: {}", e.getMessage(), e);
		}
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/application/event/NotificationEventListener.kt
package spring.web.kotlin.application.event

import org.slf4j.LoggerFactory
import org.springframework.context.event.EventListener
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import spring.web.kotlin.domain.model.entity.Notification
import spring.web.kotlin.domain.repository.MemberRepository
import spring.web.kotlin.domain.repository.NotificationRepository

@Component
@Transactional(readOnly = true)
class NotificationEventListener(
    private val memberRepository: MemberRepository,
    private val notificationRepository: NotificationRepository
) {
    private val log = LoggerFactory.getLogger(this::class.java)

    @Async
    @EventListener
    @Transactional
    fun handleNotificationEvent(notificationEvent: NotificationEvent) {
        runCatching {
            val member = memberRepository.findByIdOrNull(notificationEvent.memberId) ?: throw RuntimeException()

            notificationRepository.save(
                Notification.of(
                    member,
                    notificationEvent.title,
                    notificationEvent.message,
                    notificationEvent.url
                )
            )
        }.onFailure { e -> log.error("알림 이벤트 처리 중 예외 발생: {}", e.message, e) }
    }
}